### PR TITLE
Hibernate DB design

### PR DESCRIPTION
**Pull request by ad-47**
db_design_11_02_2020

### CODE REVIEW

**Comment by ravishetye** [comment]
you keep merging master , but the PR is against cm/versioning . It shows 50 files ! which are not related to the change.

**Comment by ad-47** [comment]
Ohh ok, it was my mistake, if cm/versioning will merge with master code then your concern will resolve.

File: backend/src/main/resources/liquibase/create-tables-changelog-1.0.xml [hunk 1]
```diff
@@ -8,7 +8,7 @@
     <changeSet author="anandJ" id="createTable-1">
         <createTable catalogName="modeldb"
                      schemaName="public"
-                     tableName="artifactStore">
+                     tableName="artifactstore">
```
Comment (evgeniilezhnin) [inline backend/src/main/resources/liquibase/create-tables-changelog-1.0.xml]: Are you sure that these name changes will not change liquibase checksum?

Comment (evgeniilezhnin) [reply backend/src/main/resources/liquibase/create-tables-changelog-1.0.xml]: When running modeldb app with a clean database set up like here:
https://github.com/pingcap/tidb/blob/master/docs/QUICKSTART.md
I get an error:
Exception in thread "main" io.grpc.StatusRuntimeException: INTERNAL: Schema-validation: missing table [artifactStore]
	at io.grpc.Status.asRuntimeException(Status.java:526)
	at io.grpc.protobuf.StatusProto.toStatusRuntimeException(StatusProto.java:52)
	at ai.verta.modeldb.utils.ModelDBHibernateUtil.getSessionFactory(ModelDBHibernateUtil.java:221)
	at ai.verta.modeldb.App.initializeServicesBaseOnDataBase(App.java:311)
	at ai.verta.modeldb.App.main(App.java:227)

Comment (evgeniilezhnin) [reply backend/src/main/resources/liquibase/create-tables-changelog-1.0.xml]: The same for the new entities

File: backend/src/main/java/ai/verta/modeldb/entities/versioning/CommitEntity.java [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+  @Column(name = "message", columnDefinition = "text")
+  private String message;
+
+  @Column(name = "date_created")
+  private Long date_created;
+
+  @Column(name = "tree_hash")
+  private String tree_hash;
+
+  @OneToOne(targetEntity = RepositoryEntity.class, cascade = CascadeType.ALL)
```
Comment (evgeniilezhnin) [inline backend/src/main/java/ai/verta/modeldb/entities/versioning/CommitEntity.java]: this looks more like many to one

File: backend/src/main/java/ai/verta/modeldb/entities/versioning/CommitEntity.java [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+  @JoinTable(
+      name = "repository_commit",
+      joinColumns = @JoinColumn(name = "commit_hash"),
+      inverseJoinColumns = @JoinColumn(name = "repository_id"))
+  private RepositoryEntity repository;
+
+  @OneToOne(targetEntity = RepositoryEntity.class, cascade = CascadeType.ALL)
```
Comment (evgeniilezhnin) [inline backend/src/main/java/ai/verta/modeldb/entities/versioning/CommitEntity.java]: this looks more like many to one

File: backend/src/main/java/ai/verta/modeldb/entities/dataset/S3DatasetBlobEntity.java [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+  public S3DatasetBlobEntity(String blobHash, S3DatasetBlob s3DatasetBlob) {
+    this.blob_hash = blobHash;
+    this.pathDatasetBlob = new PathDatasetBlobEntity(blobHash, s3DatasetBlob.getPath());
+  }
+
+  @Id
+  @Column(name = "blob_hash", nullable = false, columnDefinition = "varchar", length = 32)
+  private String blob_hash;
+
+  @OneToOne(targetEntity = PathDatasetBlobEntity.class, cascade = CascadeType.ALL)
```
Comment (evgeniilezhnin) [inline backend/src/main/java/ai/verta/modeldb/entities/dataset/S3DatasetBlobEntity.java]: I'm not 100% sure because I didn't fully understand the new design yet. But probably it's many to one too.

Comment (evgeniilezhnin) [reply backend/src/main/java/ai/verta/modeldb/entities/dataset/S3DatasetBlobEntity.java]: So it's not one to one: https://docs.google.com/document/d/1jS1h2X9A05WYV_23vvUCuMj082Ty68zGDHEgN0xPDXo/edit?disco=AAAAGN38FeI

File: protos/protos/public/modeldb/versioning/VersioningService.proto [hunk 5]
```diff
@@ -62,7 +63,7 @@ message Entity {
 }
 
 message Repository {
-    string id = 1;
+    uint64 id = 1;
```
Comment (evgeniilezhnin) [inline protos/protos/public/modeldb/versioning/VersioningService.proto]: It can't be null so how can I check if a user wants to create a new repository or update an existing one?

Comment (conradoverta) [reply protos/protos/public/modeldb/versioning/VersioningService.proto]: Make zero (default) an invalid value.

Comment (evgeniilezhnin) [reply protos/protos/public/modeldb/versioning/VersioningService.proto]: ok

File: backend/src/main/java/ai/verta/modeldb/entities/ArtifactStoreMapping.java [hunk 6]
```diff
@@ -8,7 +8,7 @@
 import javax.persistence.Table;
 
 @Entity
-@Table(name = "artifactStore")
+@Table(name = "artifact_store")
```
Comment (ravishetye) [inline backend/src/main/java/ai/verta/modeldb/entities/ArtifactStoreMapping.java]: let us extract the Table name changes into a separate PR , I can do the migration again tomorrow based on it.

File: backend/src/main/java/ai/verta/modeldb/entities/ExperimentRunEntity.java [hunk 7]
```diff
@@ -25,7 +25,7 @@
 import org.hibernate.annotations.LazyCollectionOption;
 
 @Entity
-@Table(name = "experimentRun")
+@Table(name = "experimentrun")
```
Comment (ravishetye) [inline backend/src/main/java/ai/verta/modeldb/entities/ExperimentRunEntity.java]: experiment_run

File: backend/src/main/java/ai/verta/modeldb/entities/CodeVersionEntity.java [hunk 8]
```diff
@@ -43,7 +43,7 @@ public CodeVersionEntity(String fieldType, CodeVersion codeVersion) {
   private Long date_logged;
 
   @OneToOne(cascade = CascadeType.ALL)
-  @JoinColumn(name = "git_snapshot_id")
+  @JoinColumn(name = "gitsnapshot_id")
```
Comment (ravishetye) [inline backend/src/main/java/ai/verta/modeldb/entities/CodeVersionEntity.java]: I like the previous format, unless necessary why change it?

File: backend/src/main/java/ai/verta/modeldb/entities/GitSnapshotEntity.java [hunk 9]
```diff
@@ -3,16 +3,18 @@
...
 import javax.persistence.Entity;
 import javax.persistence.GeneratedValue;
 import javax.persistence.GenerationType;
 import javax.persistence.Id;
+import javax.persistence.JoinColumn;
 import javax.persistence.Table;
 
 @Entity
-@Table(name = "gitSnapshot")
+@Table(name = "gitsnapshot")
```
Comment (ravishetye) [inline backend/src/main/java/ai/verta/modeldb/entities/GitSnapshotEntity.java]: git_snapshot

File: backend/src/main/java/ai/verta/modeldb/entities/GitSnapshotEntity.java [hunk 10]
```diff
@@ -33,6 +35,9 @@ public GitSnapshotEntity(String fieldType, GitSnapshot gitSnapshot) {
   private Long id;
 
   @ElementCollection
+  @CollectionTable(
+      name = "gitsnapshot_filepaths",
```
Comment (ravishetye) [inline backend/src/main/java/ai/verta/modeldb/entities/GitSnapshotEntity.java]: git_snapshot_file_paths

File: backend/src/main/java/ai/verta/modeldb/entities/GitSnapshotEntity.java [hunk 11]
```diff
@@ -33,6 +35,9 @@ public GitSnapshotEntity(String fieldType, GitSnapshot gitSnapshot) {
...
+      joinColumns = @JoinColumn(name = "gitsnapshot_id"))
```
Comment (ravishetye) [inline backend/src/main/java/ai/verta/modeldb/entities/GitSnapshotEntity.java]: git_snapshot_id

File: backend/src/main/java/ai/verta/modeldb/entities/TagsMapping.java [hunk 12]
```diff
@@ -11,7 +11,7 @@
 import javax.persistence.Table;
 
 @Entity
-@Table(name = "TagMapping")
+@Table(name = "tagmapping")
```
Comment (ravishetye) [inline backend/src/main/java/ai/verta/modeldb/entities/TagsMapping.java]: tag_mapping

File: backend/src/main/resources/liquibase/create-tables-changelog-1.0.xml [hunk 13]
```diff
@@ -8,7 +8,7 @@
     <changeSet author="anandJ" id="createTable-1">
```
Comment (ravishetye) [inline backend/src/main/resources/liquibase/create-tables-changelog-1.0.xml]: extract these to a different PR

File: backend/src/main/resources/liquibase/create_indexes.sql [hunk 14]
```diff
@@ -7,29 +7,29 @@ CREATE  INDEX kv_er_id on keyvalue (experiment_run_id);
 CREATE  INDEX a_e_id on artifact (experiment_id);
 CREATE  INDEX a_er_id on artifact (experiment_run_id);
 CREATE  INDEX a_p_id on artifact (project_id);
-CREATE  INDEX t_e_id on TagMapping (experiment_id);
```
Comment (ravishetye) [inline backend/src/main/resources/liquibase/create_indexes.sql]: extract these to new PR

File: backend/src/main/resources/liquibase/db-changelog-1.0.xml [hunk 15]
```diff
@@ -6,6 +6,7 @@
   logicalFilePath="src/main/resources/liquibase/db-changelog-1.0.xml">
 
 	<include file="src/main/resources/liquibase/create-tables-changelog-1.0.xml"/>
+	<include file="src/main/resources/liquibase/create-versioning-tables-changelog.xml"/>
```
Comment (ravishetye) [inline backend/src/main/resources/liquibase/db-changelog-1.0.xml]: Can the include tag be placed lower in middle of the changeset?
then I would prefer that to get a sense of DB changes from going from top to down.

File: backend/src/main/resources/liquibase/hibernate-base-db-schema.sql [hunk 16]
```diff
@@ -695,3 +695,678 @@ alter table TagMapping add constraint FK6don0cq34v6j75a6ddvfft5mq foreign key (e
 alter table TagMapping add constraint FK7qfxoshi1y5vnn42yvxklr7hm foreign key (experiment_run_id) references experimentRun;
 alter table TagMapping add constraint FKht136fxcojwibhx48k7iolc1u foreign key (project_id) references project;
 alter table user_comment add constraint FK8run8dgvadrrwcwe5xpdscynm foreign key (comment_id) references comment;
+alter table artifact drop constraint FKb1510ajwbgmwqk9mk7s5madq8;
```
Comment (ravishetye) [inline backend/src/main/resources/liquibase/hibernate-base-db-schema.sql]: This file is not used anywhere right?

File: protos/protos/public/modeldb/versioning/VersioningService.proto [hunk 17]
```diff
@@ -10,9 +10,10 @@ import "protos/public/modeldb/versioning/Dataset.proto";
```
Comment (ravishetye) [inline protos/protos/public/modeldb/versioning/VersioningService.proto]: I do not expect any protos change as part of this PR , please revert these if stray andcomment on other protos PR if changes required OR file a separate protos PR

File: backend/src/main/resources/liquibase/hibernate-base-db-schema.sql (hunk 16 above)
Comment (ad-47) [reply backend/src/main/resources/liquibase/hibernate-base-db-schema.sql]: Yes. not it in use.

**Comment by ad-47** [comment]
Resolved review comment and separated code moved to new PR: https://github.com/VertaAI/modeldb/pull/106

**Comment by ad-47** [comment]
Can I close this PR @ravishetye ? because I have extracted all code to in two separate PR's

File: backend/src/main/resources/liquibase/create-versioning-tables-changelog.xml [hunk 18]
```diff
@@ -0,0 +1,118 @@
+<databaseChangeLog
+        xmlns="http://www.liquibase.org/xml/ns/dbchangelog"
+        xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+        xsi:schemaLocation="http://www.liquibase.org/xml/ns/dbchangelog
+                        http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-3.4.xsd"
+        logicalFilePath="src/main/resources/liquibase/create-tables-changelog-1.0.xml">
+
+    <changeSet author="anandJ" id="create-versioning-tables-1">
+        <createTable catalogName="modeldb"
```
[the remark this answers is not in the document]
Comment (ravishetye) [reply backend/src/main/resources/liquibase/create-versioning-tables-changelog.xml]: this was not resolved
